apm-sniffer/apm-sdk-plugin/dbcp-2.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/dbcp/v2/PoolingSetUrlSourceInterceptor.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.apache.skywalking.apm.plugin.dbcp.v2;
+
+import org.apache.commons.dbcp2.BasicDataSource;
+import org.apache.skywalking.apm.agent.core.logging.api.ILog;
+import org.apache.skywalking.apm.agent.core.logging.api.LogManager;
+import org.apache.skywalking.apm.agent.core.meter.MeterFactory;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.EnhancedInstance;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.InstanceMethodsAroundInterceptor;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.MethodInterceptResult;
+import org.apache.skywalking.apm.plugin.jdbc.connectionurl.parser.URLParser;
+import org.apache.skywalking.apm.plugin.jdbc.trace.ConnectionInfo;
+
+import java.lang.reflect.Method;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * {@link PoolingSetUrlSourceInterceptor} intercepted the method of DBCP set url and register metric monitor.
+ */
+public class PoolingSetUrlSourceInterceptor implements InstanceMethodsAroundInterceptor {
+    private static final String METER_NAME = "datasource";
+    private static final ILog LOGGER = LogManager.getLogger(PoolingSetUrlSourceInterceptor.class);
+    private static final Map<String, Function<BasicDataSource, Supplier<Double>>> METRIC_MAP = new HashMap<String, Function<BasicDataSource, Supplier<Double>>>();
```

Comment:
Hi @sunOnly I asked @ascrutae to help on locating what is going on about https://github.com/apache/skywalking-java/pull/96/files#r800393136

All of your plugins have issues with using the target class(in here `BasicDataSource`) in instrumentation define and also in interceptor's static field declaring. 
Then in this case, `BasicDataSource` is loaded by the agent class loader already, which is impossible to be enhanced by the plugin system. Also, we checked the plugins of druid and hikaricp plugins, they are not working too.

The right way to fix this using `javax.sql.DataSource` here for the declaring, and do typecast if you have to use it in the runtime.
Also, you should know, even you opened **BootstrapInstrumentation**, it just overrides the error, but your plugin is still not working. I think you could verify locally when you debug locally or write plugin tests.

Please change the ways we recommended, and add all required plugin tests.

Reply:
> Hi @sunOnly I asked @ascrutae to help on locating what is going on about https://github.com/apache/skywalking-java/pull/96/files#r800393136
> 
> All of your plugins have issues with using the target class(in here `BasicDataSource`) in instrumentation define and also in interceptor's static field declaring. Then in this case, `BasicDataSource` is loaded by the agent class loader already, which is impossible to be enhanced by the plugin system. Also, we checked the plugins of druid and hikaricp plugins, they are not working too.
> 
> The right way to fix this using `javax.sql.DataSource` here for the declaring, and do typecast if you have to use it in the runtime. Also, you should know, even you opened **BootstrapInstrumentation**, it just overrides the error, but your plugin is still not working. I think you could verify locally when you debug locally or write plugin tests.
> 
> Please change the ways we recommended, and add all required plugin tests.

received